src/io_callbacks.zig

```diff
@@ -0,0 +1,463 @@
+                                os.ENAMETOOLONG => error.NameTooLong,
+                                os.ENFILE => error.SystemFdQuotaExceeded,
+                                os.ENODEV => error.NoDevice,
+                                os.ENOENT => error.FileNotFound,
+                                os.ENOMEM => error.SystemResources,
+                                os.ENOSPC => error.NoSpaceLeft,
+                                os.ENOTDIR => error.NotDir,
+                                os.EOPNOTSUPP => error.FileLocksNotSupported,
+                                os.EOVERFLOW => error.FileTooBig,
+                                os.EPERM => error.AccessDenied,
+                                os.EWOULDBLOCK => error.WouldBlock,
+                                else => |errno| os.unexpectedErrno(@intCast(usize, errno)),
+                            };
+                        } else {
+                            break :result @intCast(os.fd_t, completion.result);
+                        }
+                    };
+                    completion.callback(completion.context, completion, &result);
+                },
+            }
+        }
+    };
+
+    /// This union encodes the set of operations supported as well as their arguments.
+    const Operation = union(enum) {
+        accept: struct {
+            socket: os.fd_t,
+            address: os.sockaddr,
+            address_size: os.socklen_t,
+            flags: u32,
+        },
+        close: struct {
+            fd: os.fd_t,
+        },
+        openat: struct {
+            fd: os.fd_t,
+            path: [*:0]const u8,
+            flags: u32,
+            mode: os.mode_t,
+        },
+    };
+
+    pub const AcceptError = error{
+        WouldBlock,
+        FileDescriptorInvalid,
+        ConnectionAborted,
+        SocketNotListening,
+        ProcessFdQuotaExceeded,
+        SystemFdQuotaExceeded,
+        SystemResources,
+        FileDescriptorNotASocket,
+        OperationNotSupported,
+        PermissionDenied,
+        ProtocolFailure,
+    } || os.UnexpectedError;
+
+    pub fn accept(
+        self: *IO,
+        completion: *Completion,
+        comptime Context: type,
+        context: Context,
+        comptime callback: fn (context: Context, completion: *Completion, result: AcceptError!os.socket_t) void,
+        socket: os.socket_t,
+        address: os.sockaddr,
+        address_size: os.socklen_t,
+        flags: u32,
+    ) void {
+        completion.* = .{
+            .io = self,
+            .context = context,
+            .callback = struct {
+                fn wrapper(ctx: ?*c_void, comp: *Completion, res: *const c_void) callconv(.C) void {
+                    callback(
+                        @intToPtr(Context, @ptrToInt(ctx)),
+                        comp,
+                        @intToPtr(*const AcceptError!os.socket_t, @ptrToInt(res)).*,
+                    );
+                }
+            }.wrapper,
```

Comment:
Neat trick!

Reply:
Thanks! We could also take this a bit further and use `@call(.{ .modifier = .always_inline }, ... )` to call the callback here though I don't have any performance numbers to back it up so I left it off for now.

Reply:
Good to know! We can always profile later and this is all internal within the API.